src/main.rs

```diff
@@ -187,10 +237,97 @@ fn byte_offset_to_point(file_contents: &Rope, byte_offset: usize) -> Point {
     }
 }
 
-fn tree_sitter_range_to_lsp_range(file_contents: &Rope, range: tree_sitter::Range) -> Range {
+fn byte_offset_to_position(file_contents: &Rope, byte_offset: usize) -> Position {
+    point_to_position(byte_offset_to_point(file_contents, byte_offset))
+}
+
+fn byte_offset_range_to_lsp_range(file_contents: &Rope, range: ops::Range<usize>) -> Range {
     Range {
-        start: point_to_position(byte_offset_to_point(file_contents, range.start_byte)),
-        end: point_to_position(byte_offset_to_point(file_contents, range.end_byte)),
+        start: byte_offset_to_position(file_contents, range.start),
+        end: byte_offset_to_position(file_contents, range.end),
+    }
+}
+
+fn tree_sitter_range_to_lsp_range(file_contents: &Rope, range: tree_sitter::Range) -> Range {
+    byte_offset_range_to_lsp_range(file_contents, range.start_byte..range.end_byte)
+}
+
+fn get_text_document_edits(
+    old_contents: &Rope,
+    new_contents: &Rope,
+    uri: &Url,
+) -> TextDocumentEdit {
+    let old_contents_as_str: Cow<'_, str> = old_contents.clone().into();
+    let new_contents_as_str: Cow<'_, str> = new_contents.clone().into();
+    let (_, diffs) = text_diff::diff(&old_contents_as_str, &new_contents_as_str, "");
+    let mut old_bytes_seen = 0;
+    let mut just_seen_remove: Option<String> = None;
+    let mut edits: Vec<TextEdit> = Default::default();
+    for diff in diffs {
+        if just_seen_remove.is_some() && !matches!(&diff, Difference::Add(_)) {
```

Comment:
You could think of "'remove' followed immediately by 'add'" as a special case that we want to combine into a single LSP text edit

But so here this is emitting a text edit for a remove if it's _not_ immediately followed by an add